### ISSUE TITLE
how to set a watchpoint in vscode dlv? As dlv support watch command.

### ISSUE THREAD

**Issue by ripplewang888** [golang/vscode-go#3509]
**Is your feature request related to a problem? Please describe.**
As gdb does, break at a watch point for when a variable value change. Delve support watch command, but I cant' execute this command 
in vscode.

**Describe the solution you'd like**
(dlv) watch  -rw b
Watchpoint b set at 0x556188
(dlv) l

how to do this in vscode? I want to set a watchpoint break.

**Describe alternatives you've considered**

**Additional context**

**Comment by ripplewang888** [comment]
![image](https://github.com/user-attachments/assets/c87a3a7e-8f7f-4454-9282-0276f324e7c1)
I think should add two menu bars when right click global variables.

**Comment by ripplewang888** [comment]
@gopherbot can you assign this issue to anyone?

**Comment by hyangah** [comment]
I guess the closest is data breakpoint. Delve DAP (the debug adapter behind the scene) is not supporting data breakpoints yet.

https://github.com/go-delve/delve/blob/a164b89df10aba455e0196222e51fa4b65e5746c/service/dap/server.go#L796-L799

Contribution to DAP (in the go-delve repo) is welcome.

**Comment by ripplewang888** [comment]
@hyangah , thanks. 
  Do you know when to support this feature?

**Comment by lwintermelon** [comment]
Since we support rr, this will be a powerful and convinient feature.

**Comment by firelizzard18** [comment]
This is not a feature that can be implemented by vscode-go, it must be implemented by the debugger, [delve](https://github.com/go-delve/delve). However, it appears unlikely that it will be implemented. If you look at delve's DAP implementation, data breakpoints are explicitly not supported and the comment above that block says "Requests that we do not plan to support".

https://github.com/go-delve/delve/blob/52a73b0ecfc50ad1eb0bcd579e2213a8ca64f5a2/service/dap/server.go#L788-L806

[This comment](https://github.com/go-delve/delve/pull/1918#discussion_r390614202) indicates that @aarzilli (one of the two maintainers of the debugger) does not think it is feasible to implement data breakpoints.

**Comment by aarzilli** [comment]
Since that comment was written we actually implemented watchpoints in delve, so now I think that data breakpoint (with limitations) could be implemented (maybe, I haven't looked at the spec)

**Comment by lwintermelon** [comment]
dlv does support [watchpoint](https://github.com/go-delve/delve/blob/master/Documentation/cli/README.md#watch), but it doesn't expose it to `DAP`, right?

I try to use `--accept-multiclient`, and connect it with `vscode-go` and `dlv connect` (which case watchpoint is available), the watchpoint of `dlv` command line client works, but it doesn't work well with `vscode-go`, it has bugs so `breakpoints` command doesn't show id of breakpoints(even breakpoints set from `dlv` command line client), making it hard to clear a breakpoint.

I still wonder whether exposing direct `dlv` command line client communication to the `DAP DEBUG CONSOLE` is a good idea, actually you can do it by `--accept-multiclient`. So we can use fuctions of `dlv` which not covered by `dlv dap` or `DAP` itself.

**Comment by firelizzard18** [comment]
> dlv does support watchpoint, but it doesn't expose it to DAP, right?

Correct. To change this, someone will need to read through the DAP specification to understand how DAP expects data breakpoints to behave, and then implement the SetDataBreakpointsRequest and DataBreakpointInfoRequest requests and add supportsDataBreakpoints to delve's DAP capabilities list.

**Comment by lwintermelon** [comment]
DataBreakpoints support for dlv DAP has been implemented upsteam, how about vscode-go？

**Comment by aarzilli** [comment]
Nothing needs to be done to vscode-go.

**Comment by firelizzard18** [comment]
Just the normal "install dlv@master or wait for a release and install dlv@latest". If you ask vscode-go to update it will install dlv@latest.

**Comment by lwintermelon** [comment]
I tried it and it works for simple cases.

I'd like to say for RR backend, the user experience can stilled be improved.

Only `rev continue` and `rev next` are acceesible by vscode UI (I know that DAP protocol only exposes these), and data breakpoints can't be customed(even don't know if it is set successfully, but in dlv cli, it will tell you). 

Like I described above, I'll connect 2 dlv client to `dlv --accept-multiclient` to overcome the shortcoming, one is vscode-go, another is dlv cli, but this way also has a shortcoming: in dlv cli, `breakpoints` subcommand can't list the id for breakpoints, even for breakpoint that be set by same dlv cli, so you can't clear a specific breakpoint! which is very inconvenient.(I don't know why...)

A simple way to improve is to allow more dlv commands in vscode-go when debugging, like `ms-vscode.cpptools`, `CodeLLDB`, `LLDB DAP`, you can call any gdb commands or lldb commands when debugging.

A better way will be GUI integration, like `undo.udb` plugins for vscode.

PS:
I'm interested in time travel debugging these days, and Microsoft has opened the API of TTD replay, thanks to the good layered architecture of delve, it won't be too hard to add a new reversible backend like RR.

**Comment by VelorumS** [comment]
> Nothing needs to be done to vscode-go.

I'd like to be able to unfold a struct in a watch window, right click on a field and do "break on value change".

**Comment by firelizzard18** [comment]
> I'd like to be able to unfold a struct in a watch window, right click on a field and do "break on value change"

@VelorumS whether features like that work is determined by the IDE (VSCode), the debugger (delve), and the [debug adapter protocol](https://microsoft.github.io/debug-adapter-protocol/). This extension (vscode-go) is only involved in starting delve; ***all*** subsequent interaction is directly between the IDE and the debugger, without any involvement from vscode-go. So, nothing needs to be done to vscode-go, or more accurately there's nothing vscode-go *can* do.
